akka-http2-support/src/main/scala/akka/http/impl/engine/http2/WrappedSslContextSPI.scala

```diff
@@ -6,49 +6,87 @@ package akka.http.impl.engine.http2
 
 import java.security.SecureRandom
 import java.{ util ⇒ ju }
-import javax.net.ssl.KeyManager
-import javax.net.ssl.SSLContext
-import javax.net.ssl.SSLContextSpi
-import javax.net.ssl.SSLEngine
-import javax.net.ssl.SSLException
-import javax.net.ssl.SSLServerSocketFactory
-import javax.net.ssl.SSLSessionContext
-import javax.net.ssl.SSLSocketFactory
-import javax.net.ssl.TrustManager
+import javax.net.ssl._
 
+import scala.collection.immutable
+import scala.collection.JavaConverters._
 import akka.http.scaladsl.HttpsConnectionContext
 import org.eclipse.jetty.alpn.ALPN
+import org.eclipse.jetty.alpn.ALPN.ServerProvider
+
+object ALPNRegistry {
+  def registerServerEngine(engine: SSLEngine)(decider: immutable.Set[String] ⇒ Option[String]): Unit =
+    ALPN.put(engine, new ServerProvider {
+      def select(protocols: ju.List[String]): String =
+        choose(protocols.asScala.toSet)
+
+      def unsupported(): Unit =
+        choose(Set.empty)
+
+      def choose(protocols: immutable.Set[String]): String =
+        try decider(protocols) match {
+          case entry @ Some(chosen) ⇒
+            if (protocols.nonEmpty && !protocols(chosen))
+              throw new SSLException(s"Tried to choose unsupported protocol. Chosen $chosen supported: [$protocols]")
+            else {
+              println(s"Chose: $chosen for engine ${engine}")
+              putSelection(engine, entry)
+              chosen
+            }
+          case None ⇒
+            println(s"Chose nothing")
+            putSelection(engine, None)
+
+            if (protocols.nonEmpty)
+              throw new SSLException(s"No protocol chosen from [$protocols]")
+            else ""
+        }
+        finally ALPN.remove(engine)
+    })
+
+  def getChosenProtocolAndRemove(engine: SSLEngine): Option[String] = synchronized {
+    val selected = selections.getOrElse(engine, None)
+    selections -= engine
+    selected
+  }
```

Comment:
Hm, doesn't this need to synchronize (I'd synchronize on selections instance btw, a bit more clear what we're protecting) as well? Synchronized itself probably fine for this use case (we benched similar in BufferPool).